### PR TITLE
refactor(hooks): refactor hooks

### PR DESCRIPTION
**Pull request by sxzz**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

File: packages/hooks/use-modal/index.ts [hunk 1]
```diff
@@ -20,25 +19,17 @@ const closeModal = (e: KeyboardEvent) => {
...
+      )
     }
-  )
+  })
 }
 
-if (!isServer) {
-  on(document, 'keydown', closeModal)
-}
+if (isClient) useEventListener(document, 'keydown', closeModal)
```
Comment (HerringtonDarkholme) [inline packages/hooks/use-modal/index.ts:35]: This is not side-effect free. Shall we use some init function instead?

**Review by HerringtonDarkholme — APPROVED** [review]
LGTM. But I would suggest break large pull requests into separate small changes.
Note this pull requests consists of several different refactors that are readily to be broken up.

**Comment by sxzz** [comment]
@HerringtonDarkholme Each refactoring corresponds to a commit. I think rebase merge is acceptable. There should be no need for a commit and a pull request.

File: packages/hooks/use-modal/index.ts (hunk 1 above)
Comment (sxzz) [reply packages/hooks/use-modal/index.ts:35]: Are there any side effects of useEventListener? I didn't get your point.

Comment (JeremyWuuuuu) [reply packages/hooks/use-modal/index.ts:35]: @HerringtonDarkholme Can you elaborate what the side effect is here?

File: packages/hooks/use-timeout/index.ts [hunk 2]
```diff
@@ -1,22 +1,16 @@
...
-  let timeoutHandle: TimeoutHandle
-
-  onBeforeUnmount(() => {
-    clearTimeout(timeoutHandle)
-  })
+  const registerTimeout = (fn: (...args: any[]) => unknown, delay: number) => {
+    cancelTimeout()
+    ;({ stop } = useTimeoutFn(fn, delay))
+  }
+  const cancelTimeout = () => stop?.()
```
Comment (JeremyWuuuuu) [inline packages/hooks/use-timeout/index.ts:10]: I think the usage here is wrong, please refer to the source code to get appropriate invoking method.

**Review by jw-foss — COMMENTED** [review]
LGTM, let's wait for @HerringtonDarkholme's input.

**Pull request merged**